lumicks/pylake/calibration.py

```diff
@@ -0,0 +1,95 @@
+from collections import namedtuple
+
+
+class ForceCalibration:
+    """Calibration handling
+
+    Parameters
+    ----------
+    calibration: named tuple containing
+        time_field : name of the field used for time
+        items : list of dictionaries containing the raw calibration attribute data
+    """
+    def __init__(self, calibration):
+        if not calibration:
+            raise ValueError("Missing argument calibration")
+
+        if not hasattr(calibration, "items"):
+            raise TypeError("Calibration structure is missing items field")
+
+        if not hasattr(calibration, "time_field"):
+            raise TypeError("Calibration structure is missing time_field field")
+
+        self._calibration = calibration
+
+    """filter calibration data based on time stamp range [ns]"""
+    @staticmethod
+    def _filter_calibration(calibration, start, stop):
+        assert calibration
+
+        if len(calibration.items) == 0:
+            return []
+
+        def timestamp(x):
+            return x[calibration.time_field]
+
+        # Sort by time
+        calibration.items = sorted(calibration.items, key=timestamp)
+        calibration_items = [calibration.items[0]]
+        for i, v in enumerate(calibration.items):
+            ts = timestamp(v)
+            if ts <= start:
+                calibration_items[0] = v
+            elif ts < stop:
+                calibration_items.append(v)
+            else:
+                # Since the list is sorted, we can early out
+                break
```

Comment:
The previous one is only needed if start doesn't equal the calibration timestamp.

So I'd have to add something like pre and start < timestamp(calibration_items[0]), but add another check that calibration items is not empty, which gives rise to 
if pre and (not calibration_items or start < timestamp(calibration_items[0])):

In addition, there's one more corner case, if we slice with [120,120] and a calibration point exists at 120, we'd probably want 120 rather than the one before it. This code would still gives you the calibration point before that since the filter uses <. This could be fixed by using <= for the stop condition in the second filter I suppose.

What I'd end up with is this:
```
    calibration_items = [x for x in calibration.items if start <= timestamp(x) < stop]

    pre = [x for x in calibration.items if timestamp(x) <= start]

    if pre and (not calibration_items or start < timestamp(calibration_items[0])):

        calibration_items.insert(0, pre[-1])
```

Then again, it may not be desirable, since [120,120] should already return an empty slice. Maybe it'd make more sense to return an empty calibration struct to signal that the slice will be empty. In the timeseries case that actually already happens since the slicing there is not lazily evaluated. What do you think?

Personally, the for loop is easier to read for me, but I should probably get used to writing more in terms of list comprehensions; since it's more in style with python. I will get used to this and then this'll probably feel more natural.

Reply:
Indeed, you are right, my previous code does not quite do the right thing. Modifying the conditions a bit would resolve this:
```python
# before
results = [x for x in items if start <= timestamp(x) < stop]
pre = [x for x in items if timestamp(x) < start]

# after
results = [x for x in items if start < timestamp(x) < stop]
pre = [x for x in items if timestamp(x) <= start]
```
There should be no need for the extra conditions in this case. But anyway, I don't want to push for the comprehensions since they can be a bit of a foreign concept to start with and also not really more efficient in this case. But in the long run, I do really like them because, once used to them, they make the logic clearer and more composable.

Reply:
Yep. That's a lot cleaner. I must have been sleepy when I wrote those ugly extra conditions.

I'll update it, thanks! 👍